cpp/arcticdb/util/key_utils.hpp

```diff
@@ -163,36 +163,44 @@ inline ankerl::unordered_dense::set<AtomKey> recurse_index_keys(
     ankerl::unordered_dense::set<AtomKey> res;
     ankerl::unordered_dense::set<AtomKeyPacked> res_packed;
     for (const auto& index_key: keys) {
-        if (index_key.type() == KeyType::MULTI_KEY) {
-            // recurse_index_key includes the input key in the returned set, remove this here
-            auto sub_keys = recurse_index_key(store, index_key);
-            sub_keys.erase(index_key);
-            for (auto&& key: sub_keys) {
-                res.emplace(std::move(key));
-            }
-        } else if (index_key.type() == KeyType::TABLE_INDEX) {
-            KeySegment key_segment(store->read_sync(index_key, opts).second, SymbolStructure::SAME);
-            auto data_keys = key_segment.materialise();
-            util::variant_match(
+        try {
+            if (index_key.type() == KeyType::MULTI_KEY) {
+                // recurse_index_key includes the input key in the returned set, remove this here
+                auto sub_keys = recurse_index_key(store, index_key);
+                sub_keys.erase(index_key);
+                for (auto &&key : sub_keys) {
+                    res.emplace(std::move(key));
+                }
+            } else if (index_key.type() == KeyType::TABLE_INDEX) {
```

Comment:
I considered pulling out the different behaviours based on type to functions as this is getting quite messy, but the increased indirection made things harder to follow rather than easier.